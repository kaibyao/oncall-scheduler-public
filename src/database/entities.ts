import type { GhostEngPod, OncallRotationName } from '../schedule/schedule.types.js';

export interface OncallScheduleEntity {
  id: number;
  date: string;
  rotation: OncallRotationName;
  engineer_email: string;
}

export type OncallScheduleOverrideEntity = OncallScheduleEntity;

export interface UserEntity {
  email: string;
  name: string;
  slack_user_id: string | null;
  notion_person_id: string | null;
  rotation: string;
  pod: GhostEngPod;
  created_at?: string;
  updated_at?: string;
  deleted_at?: string;
}

export type Upsertable<T> = Omit<T, 'id' | 'created_at' | 'updated_at' | 'deleted_at'>;
