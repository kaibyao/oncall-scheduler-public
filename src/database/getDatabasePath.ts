import path from 'path';
import { IS_PRODUCTION } from '../config.js';

export function getDatabasePath(): string {
  if (IS_PRODUCTION) {
    return '/mnt/efs/oncall_schedule.db';
  } else {
    // Use the same path as db.ts for consistency
    const __dirname = path.dirname(new URL(import.meta.url).pathname);
    return path.join(__dirname, '../../database/oncall_schedule.db');
  }
}
