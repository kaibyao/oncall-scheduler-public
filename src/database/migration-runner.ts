/**
 * Database migration runner - isolated from the main database instance
 * This module contains the core migration logic that can be used with any database instance
 */

import * as fs from 'fs';
import * as path from 'path';
import type Database from 'better-sqlite3';
import { Logger } from '../logger.js';

const logger = new Logger('migration-runner');

/**
 * Runs all database migrations on the provided database instance
 * @param db - Database instance to run migrations on
 * @param migrationsPath - Optional path to migrations directory (defaults to 'migrations' in cwd)
 */
export function runMigrations(db: Database.Database, migrationsPath?: string): void {
  // Create migrations tracking table if it doesn't exist
  db.exec(`
    CREATE TABLE IF NOT EXISTS schema_migrations (
      version TEXT PRIMARY KEY,
      applied_at DATETIME DEFAULT CURRENT_TIMESTAMP
    )
  `);

  const migrationsDir = migrationsPath || path.join(process.cwd(), 'migrations');

  // Get all .sql files from migrations directory
  const migrationFiles = fs
    .readdirSync(migrationsDir)
    .filter((file) => file.endsWith('.sql'))
    .sort(); // Ensure consistent ordering

  const appliedMigrations = new Set(
    (db.prepare('SELECT version FROM schema_migrations').all() as { version: string }[]).map((row) => row.version),
  );

  logger.info(`Found ${migrationFiles.length} migration files`);

  for (const file of migrationFiles) {
    const version = path.basename(file, '.sql');

    if (appliedMigrations.has(version)) {
      logger.info(`Skipping ${file} (already applied)`);
      continue;
    }

    logger.info(`Applying migration: ${file}`);

    const migrationPath = path.join(migrationsDir, file);
    const sql = fs.readFileSync(migrationPath, 'utf8');

    try {
      // Execute the migration in a transaction
      db.transaction(() => {
        db.exec(sql);
        db.prepare('INSERT INTO schema_migrations (version) VALUES (?)').run(version);
      })();

      logger.info(`✓ Applied ${file}`);
    } catch (error) {
      logger.error(`✗ Failed to apply ${file}:`, error);
      throw error;
    }
  }

  logger.info('All migrations completed');
}
