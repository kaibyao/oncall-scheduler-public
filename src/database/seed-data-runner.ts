import Database from 'better-sqlite3';
import fs from 'fs';
import path from 'path';
import { Logger } from '../logger.js';

const logger = new Logger('seed-data-runner');

/**
 * Runs seed data files from the seed-data directory, tracking which files have been applied
 * to avoid re-executing them. Similar to migration runner but for seed data.
 */
export function runSeedData(db: Database.Database, seedDataPath?: string): void {
  // Create seed data tracking table if it doesn't exist
  db.exec(`
    CREATE TABLE IF NOT EXISTS seed_data_applied (
      name TEXT PRIMARY KEY,
      applied_at DATETIME DEFAULT CURRENT_TIMESTAMP
    )
  `);

  const seedDataDir = seedDataPath || path.join(process.cwd(), 'seed-data');

  // Get all .sql files from seed-data directory
  const seedFiles = fs
    .readdirSync(seedDataDir)
    .filter((file) => file.endsWith('.sql'))
    .sort(); // Ensure consistent ordering

  const appliedSeedData = new Set(
    (db.prepare('SELECT name FROM seed_data_applied').all() as { name: string }[]).map((row) => row.name),
  );

  logger.info(`Seed data already applied`, { appliedSeedData });

  logger.info(`Found ${seedFiles.length} seed data files`);

  for (const file of seedFiles) {
    const name = path.basename(file, '.sql');

    if (appliedSeedData.has(name)) {
      logger.info(`Skipping ${file} (already applied)`);
      continue;
    }

    logger.info(`Applying seed data: ${file}`);

    const seedFilePath = path.join(seedDataDir, file);
    const sql = fs.readFileSync(seedFilePath, 'utf8');

    try {
      // Execute the seed data in a transaction
      db.transaction(() => {
        db.exec(sql);
        db.prepare('INSERT INTO seed_data_applied (name) VALUES (?)').run(name);
      })();

      logger.info(`✓ Applied ${file}`);
    } catch (error) {
      logger.error(`✗ Failed to apply ${file}:`, error);
      throw error;
    }
  }

  logger.info('All seed data completed');
}
