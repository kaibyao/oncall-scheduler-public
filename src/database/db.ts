import Database from 'better-sqlite3';
import fs from 'fs';
import { Logger } from '../logger.js';
import { getDatabasePath } from './getDatabasePath.js';

const logger = new Logger('db');

const dbPath = getDatabasePath();

logger.info('Connecting to database at: ', dbPath);
if (!fs.existsSync(dbPath)) {
  logger.info('Database file does not exist, creating...');
  fs.writeFileSync(dbPath, '');
  logger.info('Database file created at: ', dbPath);
} else {
  logger.info('Database file exists at: ', dbPath);
}

const db = new Database(dbPath);
db.pragma('journal_mode = WAL');
db.pragma('foreign_keys = ON');

export default db;
