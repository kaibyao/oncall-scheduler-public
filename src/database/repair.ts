import * as fs from 'fs';
import * as path from 'path';
import { execSync } from 'child_process';
import { Logger } from '../logger.js';
import Database from 'better-sqlite3';
import { getDatabasePath } from './getDatabasePath.js';

const logger = new Logger('db-repair');

export interface RepairResult {
  success: boolean;
  backupPath?: string;
  recoveredTables?: number;
  recoveredRows?: number;
  errors?: string[];
  warnings?: string[];
}

/**
 * Repairs a corrupt SQLite database using the SQLite recovery tool
 */
export function repairDatabase(): RepairResult {
  const dbPath = getDatabasePath();
  const result: RepairResult = {
    success: false,
    errors: [],
    warnings: [],
  };

  logger.info('Starting database repair process for:', dbPath);

  try {
    // Check if database file exists
    if (!fs.existsSync(dbPath)) {
      result.errors!.push('Database file does not exist');
      logger.error('Database file does not exist:', dbPath);
      return result;
    }

    // Create backup filename with today's date
    const today = new Date().toISOString().split('T')[0]; // YYYY-MM-DD format
    const backupPath = path.join(path.dirname(dbPath), `${today}_corrupt_db_before_repair.db`);
    result.backupPath = backupPath;

    logger.info('Creating backup at:', backupPath);

    // Step 1: Move the existing DB file to backup
    fs.copyFileSync(dbPath, backupPath);
    logger.info('✓ Database backed up successfully');

    // Step 2: Generate recovery SQL
    const recoveryPath = path.join(path.dirname(dbPath), 'recovery.sql');
    logger.info('Generating recovery SQL...');

    try {
      // Use sqlite3 .recover command to extract recoverable data
      const recoverCommand = `sqlite3 "${backupPath}" .recover`;
      const recoverySQL = execSync(recoverCommand, {
        encoding: 'utf8',
        maxBuffer: 50 * 1024 * 1024, // 50MB buffer for large databases
      });

      // Write recovery SQL to file
      fs.writeFileSync(recoveryPath, recoverySQL);
      logger.info('✓ Recovery SQL generated');

      // Check if recovery SQL is not empty
      if (!recoverySQL.trim()) {
        result.warnings!.push('Recovery SQL is empty - database may be severely corrupted');
        logger.warn('Warning: Recovery SQL is empty');
      }
    } catch (error) {
      result.errors!.push(`Failed to generate recovery SQL: ${error}`);
      logger.error('Failed to generate recovery SQL:', error);
      return result;
    }

    // Step 3: Create new database from recovery SQL
    logger.info('Creating new database from recovery SQL...');

    try {
      // Remove the original corrupted database
      if (fs.existsSync(dbPath)) {
        fs.unlinkSync(dbPath);
      }

      // Create new database by running recovery SQL
      const restoreCommand = `sqlite3 "${dbPath}" < "${recoveryPath}"`;
      execSync(restoreCommand, {
        encoding: 'utf8',
        stdio: 'pipe', // Suppress output
      });

      logger.info('✓ New database created from recovery SQL');
    } catch (error) {
      result.errors!.push(`Failed to restore database: ${error}`);
      logger.error('Failed to restore database:', error);

      // Try to restore backup if restoration failed
      if (fs.existsSync(backupPath)) {
        logger.info('Attempting to restore from backup...');
        fs.copyFileSync(backupPath, dbPath);
        result.warnings!.push('Restoration failed, original database restored from backup');
      }

      return result;
    }

    // Step 4: Validate the recovered database
    logger.info('Validating recovered database...');

    try {
      const recoveredDb = new Database(dbPath, { readonly: true });

      // Check if essential tables exist
      const tables = recoveredDb.prepare("SELECT name FROM sqlite_master WHERE type='table'").all() as {
        name: string;
      }[];
      const tableNames = tables.map((t) => t.name);

      result.recoveredTables = tables.length;
      logger.info(`✓ Recovered ${tables.length} tables:`, tableNames.join(', '));

      // Count total rows across all tables
      let totalRows = 0;
      for (const table of tables) {
        if (table.name !== 'sqlite_sequence') {
          try {
            const count = recoveredDb.prepare(`SELECT COUNT(*) as count FROM "${table.name}"`).get() as {
              count: number;
            };
            totalRows += count.count;
          } catch (error) {
            logger.warn(`Could not count rows in table ${table.name}:`, error);
          }
        }
      }

      result.recoveredRows = totalRows;
      logger.info(`✓ Total recovered rows: ${totalRows}`);

      // Check for schema_migrations table (important for this application)
      if (!tableNames.includes('schema_migrations')) {
        result.warnings!.push('schema_migrations table not found - migrations may need to be re-run');
        logger.warn('Warning: schema_migrations table not found');
      }

      recoveredDb.close();
    } catch (error) {
      result.errors!.push(`Failed to validate recovered database: ${error}`);
      logger.error('Failed to validate recovered database:', error);
      return result;
    }

    // Step 5: Clean up temporary files
    try {
      if (fs.existsSync(recoveryPath)) {
        fs.unlinkSync(recoveryPath);
        logger.info('✓ Cleaned up temporary recovery file');
      }
    } catch (error) {
      result.warnings!.push(`Could not clean up temporary file: ${error}`);
      logger.warn('Could not clean up temporary file:', error);
    }

    result.success = true;
    logger.info('✓ Database repair completed successfully');

    return result;
  } catch (error) {
    result.errors!.push(`Unexpected error during repair: ${error}`);
    logger.error('Unexpected error during repair:', error);
    return result;
  }
}

/**
 * Checks if the database appears to be corrupt by attempting basic operations
 */
export function checkDatabaseHealth(dbPath?: string): { healthy: boolean; error?: string } {
  const targetPath = dbPath || getDatabasePath();

  try {
    if (!fs.existsSync(targetPath)) {
      return { healthy: false, error: 'Database file does not exist' };
    }

    // Try to open database and perform a simple query
    const db = new Database(targetPath, { readonly: true });

    try {
      // Try to read sqlite_master table
      db.prepare("SELECT name FROM sqlite_master WHERE type='table' LIMIT 1").get();
      db.close();
      return { healthy: true };
    } catch (error) {
      db.close();
      return { healthy: false, error: `Database corruption detected: ${error}` };
    }
  } catch (error) {
    return { healthy: false, error: `Cannot access database: ${error}` };
  }
}
