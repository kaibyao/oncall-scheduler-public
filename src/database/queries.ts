import db from './db.js';
import type { OncallScheduleEntity, OncallScheduleOverrideEntity, UserEntity, Upsertable } from './entities.js';
import type {
  EngineerRotationAssignment,
  EngineerRotationHours,
  OncallRotationName,
} from '../schedule/schedule.types.js';
import { Logger } from '../logger.js';
import { runMigrations } from './migration-runner.js';

const logger = new Logger('queries');

/** Returns the last date that an oncall rotation was scheduled. */
export async function getLastScheduledOncallDay(): Promise<OncallScheduleEntity | null> {
  const query = await db.prepare(`
    SELECT * FROM oncall_schedule
    ORDER BY date DESC
    LIMIT 1
  `);
  const results = (await query.get()) as OncallScheduleEntity | undefined;
  return results ?? null;
}

/** Save a generated schedule to the database. */
export async function saveSchedule(schedule: Upsertable<OncallScheduleEntity>[]): Promise<void> {
  if (schedule.length === 0) {
    logger.warn('No schedule entries to save');
    return;
  }

  // Insert all schedule entries
  for (const scheduleEntry of schedule) {
    const query = db.prepare(`
        INSERT INTO oncall_schedule (date, rotation, engineer_email) VALUES (
          ?, ?, ?
        )
        ON CONFLICT (date, rotation) DO UPDATE SET
          engineer_email = EXCLUDED.engineer_email
      `);
    query.run(scheduleEntry.date, scheduleEntry.rotation, scheduleEntry.engineer_email);
  }
}

/** Get workload history for the past N days to help with fair distribution. */
export function getWorkloadHistory(daysBack: number): EngineerRotationAssignment[] {
  /*service='oncall.data',function='getWorkloadHistory'*/

  const query = db.prepare(`
    SELECT
      os.date,
      os.rotation,
      os.engineer_email,
      COALESCE(u.name, os.engineer_email) as engineer_name
    FROM oncall_schedule os
    LEFT JOIN users u ON os.engineer_email = u.email
    WHERE os.date >= date('now', '-' || ? || ' days')
    ORDER BY os.date DESC
  `);

  return query.all(daysBack) as EngineerRotationAssignment[];
}

/** Get the total hours worked by each engineer for each rotation in the past N days. */
export function getWorkloadHistoryHoursByEngineerRotation(daysBack: number): EngineerRotationHours[] {
  const query = db.prepare(`
    WITH oncall_schedule_with_hours AS (
      SELECT
        *,
        CASE
          WHEN rotation = 'AM' THEN 3
          WHEN rotation = 'Core' THEN 6
          WHEN rotation = 'PM' THEN 3
        END AS hours
      FROM oncall_schedule
      WHERE date >= date('now', '-' || ? || ' days')
    )
    SELECT
      engineer_email,
      rotation,
      SUM(hours) as total_hours
    FROM
      oncall_schedule_with_hours
    GROUP BY
      engineer_email,
      rotation
    ORDER BY
      SUM(hours)
  `);

  return query.all(daysBack) as EngineerRotationHours[];
}

/** Get the current assignments for the next 7 days. */
export function getCurrentAssignments(): EngineerRotationAssignment[] {
  const query = db.prepare(`
    SELECT
      date,
      rotation,
      engineer_email
    FROM
      oncall_schedule
    WHERE
      date >= date('now')
      AND date < date('now', '+7 days')
    ORDER BY date ASC
  `);

  return query.all() as EngineerRotationAssignment[];
}

/** Get the current overrides for the next 7 days. */
export function getCurrentOverrides(): EngineerRotationAssignment[] {
  const query = db.prepare(`
      SELECT
        oso.date,
        oso.rotation,
        oso.engineer_email,
        COALESCE(u.name, oso.engineer_email) as engineer_name
      FROM
        oncall_schedule_overrides oso
        LEFT JOIN users u ON oso.engineer_email = u.email
      WHERE
        oso.date >= date('now')
        AND oso.date < date('now', '+7 days')
      ORDER BY oso.date ASC
    `);

  return query.all() as EngineerRotationAssignment[];
}

/**
 * Gets all schedule overrides regardless of date
 */
export function getAllOverrides(): EngineerRotationAssignment[] {
  const query = db.prepare(`
      SELECT
        oso.date,
        oso.rotation,
        oso.engineer_email,
        COALESCE(u.name, oso.engineer_email) as engineer_name
      FROM
        oncall_schedule_overrides oso
        LEFT JOIN users u ON oso.engineer_email = u.email
      ORDER BY oso.date ASC
    `);

  return query.all() as EngineerRotationAssignment[];
}

/** Upsert an override for a given date and rotation. */
export async function upsertOverrides(overrides: Upsertable<OncallScheduleOverrideEntity>[]): Promise<void> {
  for (const overrideToUpsert of overrides) {
    const query = db.prepare(`
        INSERT INTO oncall_schedule_overrides (date, rotation, engineer_email) VALUES (
          ?, ?, ?
        )
        ON CONFLICT (date, rotation) DO UPDATE SET
          engineer_email = EXCLUDED.engineer_email
      `);

    query.run(overrideToUpsert.date, overrideToUpsert.rotation, overrideToUpsert.engineer_email);
  }
}

/** Delete an override for a given date and rotation. */
export async function deleteOverride({
  startDate,
  endDate,
  rotation,
}: {
  startDate: string;
  endDate: string;
  rotation: OncallRotationName;
}): Promise<void> {
  const query = db.prepare(`
      DELETE FROM oncall_schedule_overrides
      WHERE date >= ?
        AND date <= ?
        AND rotation = ?
    `);
  query.run(startDate, endDate, rotation);
}

/** Get a user by email address. */
export function getUserByEmail(email: string): UserEntity | null {
  const query = db.prepare(`
    SELECT * FROM users
    WHERE email = ?
  `);

  const result = query.get(email) as UserEntity | undefined;
  return result ?? null;
}

/** Get all users for a specific rotation. */
export function getUsersByRotation(rotation: OncallRotationName): UserEntity[] {
  const query = db.prepare(`
    SELECT * FROM users
    WHERE rotation = ?
  `);

  return query.all(rotation) as UserEntity[];
}

export function getAllUsers(): UserEntity[] {
  const query = db.prepare(`
    SELECT *
    FROM users
    ORDER BY email
  `);

  return query.all() as UserEntity[];
}

/** Update user fields. */
export function updateUser(email: string, data: Partial<UserEntity>): void {
  const updateFields: string[] = [];
  const values: (string | null)[] = [];

  // Build dynamic UPDATE query based on provided fields
  Object.entries(data).forEach(([key, value]) => {
    if (key !== 'email' && key !== 'created_at') {
      // Don't allow updating email or created_at
      updateFields.push(`${key} = ?`);
      values.push(value);
    }
  });

  if (updateFields.length === 0) {
    logger.warn('No valid fields to update for user', { email });
    return;
  }

  // Always update the updated_at timestamp
  updateFields.push('updated_at = CURRENT_TIMESTAMP');
  values.push(email); // Add email for WHERE clause

  const query = db.prepare(`
    UPDATE users
    SET ${updateFields.join(', ')}
    WHERE email = ?
  `);

  query.run(...values);
}

/** Upsert user (insert or update). */
export function upsertUser(user: Upsertable<UserEntity>): void {
  const query = db.prepare(`
    INSERT INTO users (email, name, slack_user_id, notion_person_id, rotation, pod)
    VALUES (?, ?, ?, ?, ?, ?)
    ON CONFLICT (email) DO UPDATE SET
      name = EXCLUDED.name,
      slack_user_id = EXCLUDED.slack_user_id,
      notion_person_id = EXCLUDED.notion_person_id,
      rotation = EXCLUDED.rotation,
      pod = EXCLUDED.pod,
      updated_at = CURRENT_TIMESTAMP
  `);

  query.run(user.email, user.name, user.slack_user_id, user.notion_person_id, user.rotation, user.pod);
}

export function findEngineersBeingReplaced(dates: string[], rotation: string): string[] {
  const placeholders = dates.map(() => '?').join(',');

  // First check if there are existing overrides for these dates/rotation
  const overrideQuery = db.prepare(`
    SELECT DISTINCT engineer_email, date
    FROM oncall_schedule_overrides
    WHERE date IN (${placeholders})
      AND rotation = ?
  `);

  const existingOverrides = overrideQuery.all(...dates, rotation) as { engineer_email: string; date: string }[];
  const overriddenDates = new Set(existingOverrides.map((o) => o.date));

  // For dates that already have overrides, use those engineers
  const replacedEngineers: string[] = existingOverrides.map((o) => o.engineer_email);

  // For dates without existing overrides, check the original schedule
  const remainingDates = dates.filter((date) => !overriddenDates.has(date));

  if (remainingDates.length > 0) {
    const remainingPlaceholders = remainingDates.map(() => '?').join(',');
    const scheduleQuery = db.prepare(`
      SELECT DISTINCT engineer_email
      FROM oncall_schedule
      WHERE date IN (${remainingPlaceholders})
        AND rotation = ?
    `);

    const scheduledEngineers = scheduleQuery.all(...remainingDates, rotation) as { engineer_email: string }[];
    replacedEngineers.push(...scheduledEngineers.map((s) => s.engineer_email));
  }

  // Return unique engineers being replaced
  return [...new Set(replacedEngineers)];
}

export function resetDatabase(): void {
  const dropOverridesQuery = db.prepare(`
    DROP TABLE IF EXISTS oncall_schedule_overrides;
  `);
  dropOverridesQuery.run();

  const dropScheduleQuery = db.prepare(`
    DROP TABLE IF EXISTS oncall_schedule;
  `);
  dropScheduleQuery.run();

  const dropUsersQuery = db.prepare(`
    DROP TABLE IF EXISTS users;
  `);
  dropUsersQuery.run();

  const dropMigrationsQuery = db.prepare(`
    DROP TABLE IF EXISTS schema_migrations;
  `);
  dropMigrationsQuery.run();

  const dropSeedDataAppliedQuery = db.prepare(`
    DROP TABLE IF EXISTS seed_data_applied;
  `);
  dropSeedDataAppliedQuery.run();

  runMigrations(db);
}
