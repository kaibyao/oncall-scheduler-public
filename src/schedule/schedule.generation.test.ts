import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { runScheduleGeneration } from './schedule.generation.js';
import { createTestDatabaseWithMigrations, cleanupTestDatabase, seedTestData } from '../../test/utils/database.js';
import type Database from 'better-sqlite3';
import { DateTime } from 'luxon';
import type { OncallScheduleEntity } from '../database/entities.js';
import { OncallRotationName } from './schedule.types.js';

interface TableInfo {
  name: string;
}

// Mock external integrations to prevent actual API calls
vi.mock('./schedule.notifications.js', () => ({
  updateSlackWithScheduleChanges: vi.fn().mockResolvedValue(undefined),
}));

vi.mock('../slack/slack.messages.js', () => ({
  postSlackMessage: vi.fn().mockResolvedValue(undefined),
}));

vi.mock('../slack/slack.user-groups.js', () => ({
  getSlackUserGroup: vi.fn(),
  getUserGroupMembers: vi.fn(),
  updateUserGroupMembers: vi.fn(),
}));

vi.mock('../slack/slack.users.js', () => ({
  getSlackUserIdByEmail: vi.fn(),
}));

vi.mock('../notion/notion.sync.service.js', () => ({
  NotionSyncService: vi.fn().mockImplementation(() => ({
    syncToNotion: vi.fn().mockResolvedValue({
      success: true,
      dryRun: false,
      syncStats: {
        created: 0,
        updated: 0,
        deleted: 0,
        errors: 0,
        duration: 100,
        apiCalls: 1,
      },
    }),
  })),
}));

// Store test database reference
let testDb: Database.Database;

// Mock the database module to use our test database
vi.mock('../database/db.js', () => ({
  default: new Proxy(
    {},
    {
      get(target, prop) {
        return testDb?.[prop as keyof Database.Database];
      },
    },
  ),
}));

describe('Schedule Generation (Characterization Tests)', () => {
  beforeEach(() => {
    // Create fresh test database for each test
    testDb = createTestDatabaseWithMigrations();

    // Seed user data for schedule generation
    const seedUsers = testDb.prepare(`
      INSERT INTO users (email, name, slack_user_id, notion_person_id, rotation, pod)
      VALUES (?, ?, ?, ?, ?, ?)
    `);

    const users = [
      // AM rotation users
      ['alex@company.com', 'Alex Porras', null, null, 'AM', 'Swayze'],
      ['dave.cowart@company.com', 'Dave Cowart', null, null, 'AM', 'Blinky'],
      ['jason@company.com', 'Jason Sautieres', null, null, 'AM', 'Swayze'],
      ['jeffrey@company.com', 'Jeffrey Sun', null, null, 'AM', 'Zero'],
      ['jose.diaz@company.com', 'Jose Diaz', null, null, 'AM', 'Zero'],
      ['kai.yao@company.com', 'Kai Yao', null, null, 'AM', 'Blinky'],
      // PM rotation users
      ['bee.mcbride@company.com', 'Bee Mcbride', null, null, 'PM', 'Swayze'],
      ['cat.lee@company.com', 'Cat Lee', null, null, 'PM', 'Swayze'],
      ['eng.02@company.com', 'Engineer 2', null, null, 'PM', 'Zero'],
      ['frances.jurek@company.com', 'Frances Jurek', null, null, 'PM', 'Zero'],
      ['hamp.goodwin@company.com', 'Hamp Goodwin', null, null, 'PM', 'Blinky'],
      ['kelwen@company.com', 'Kelwen Peng', null, null, 'PM', 'Blinky'],
      ['manasa.tipparam@company.com', 'Manasa Tipparam', null, null, 'PM', 'Zero'],
      ['eng.03@company.com', 'Engineer 3', null, null, 'PM', 'Zero'],
      ['ryan.oillataguerre@company.com', 'Ryan Oillataguerre', null, null, 'PM', 'Blinky'],
      ['eng.01@company.com', 'Engineer 1', null, null, 'PM', 'Zero'],
    ];

    users.forEach((user) => seedUsers.run(...user));

    // Mock current time to make tests deterministic
    const mockDate = DateTime.fromISO('2025-07-26T12:00:00', { zone: 'America/Los_Angeles' });
    vi.useFakeTimers();
    vi.setSystemTime(mockDate.toJSDate());
  });

  afterEach(() => {
    cleanupTestDatabase(testDb);
    vi.useRealTimers();
    vi.clearAllMocks();
  });

  describe('runScheduleGeneration', () => {
    it('should complete the full generation workflow without errors', async () => {
      // Seed some historical data to make the algorithm more realistic
      seedTestData(testDb, {
        schedules: [
          {
            date: '2025-07-20',
            rotation: OncallRotationName.AM,
            engineer_email: 'engineer1@ghost.org',
          },
          {
            date: '2025-07-20',
            rotation: OncallRotationName.Core,
            engineer_email: 'engineer2@ghost.org',
          },
          {
            date: '2025-07-20',
            rotation: OncallRotationName.PM,
            engineer_email: 'engineer3@ghost.org',
          },
        ],
      });

      // This tests the integration of generation + notifications using real algorithm
      await expect(runScheduleGeneration()).resolves.not.toThrow();

      // Verify that notifications were attempted to be sent
      const { updateSlackWithScheduleChanges } = await import('./schedule.notifications.js');
      expect(updateSlackWithScheduleChanges).toHaveBeenCalledOnce();
    });

    it('should handle empty database state', async () => {
      // Test with no historical data - should still complete without errors
      await expect(runScheduleGeneration()).resolves.not.toThrow();

      // Verify that notifications were attempted to be sent even with empty state
      const { updateSlackWithScheduleChanges } = await import('./schedule.notifications.js');
      expect(updateSlackWithScheduleChanges).toHaveBeenCalled();
    });

    it('should capture current generation algorithm behavior in snapshot', async () => {
      // Seed consistent test data for reproducible results
      seedTestData(testDb, {
        schedules: [
          {
            date: '2025-07-20',
            rotation: OncallRotationName.AM,
            engineer_email: 'test1@ghost.org',
          },
          {
            date: '2025-07-21',
            rotation: OncallRotationName.Core,
            engineer_email: 'test2@ghost.org',
          },
          {
            date: '2025-07-22',
            rotation: OncallRotationName.PM,
            engineer_email: 'test3@ghost.org',
          },
        ],
      });

      await runScheduleGeneration();

      // Capture the complete schedule state to detect algorithm changes
      const generatedSchedule = testDb
        .prepare('SELECT date, rotation, engineer_email FROM oncall_schedule WHERE date > ? ORDER BY date, rotation')
        .all('2025-07-22'); // Only check newly generated entries

      // This snapshot will fail if the generation algorithm changes
      expect(generatedSchedule).toMatchSnapshot('generation-algorithm-baseline');
    });
  });

  describe('Database Integration', () => {
    it('should save generated schedule to database after running generation', async () => {
      await runScheduleGeneration();

      // Verify data was saved to database
      const savedSchedule = testDb.prepare('SELECT * FROM oncall_schedule ORDER BY date, rotation').all();

      expect(savedSchedule.length).toBeGreaterThan(0);

      // Each saved record should have the expected structure
      savedSchedule.forEach((record) => {
        expect(record).toHaveProperty('id');
        expect(record).toHaveProperty('date');
        expect(record).toHaveProperty('rotation');
        expect(record).toHaveProperty('engineer_email');
      });

      // Verify schedule extends into the future (14 day lookahead)
      const futureDates = (savedSchedule as OncallScheduleEntity[]).filter(
        (record) => new Date(record.date) > new Date('2025-07-26'),
      );
      expect(futureDates.length).toBeGreaterThan(0);
    });

    it('should use real database operations for characterization testing', async () => {
      // Verify we're using a real database with migrations
      const tables = testDb.prepare("SELECT name FROM sqlite_master WHERE type='table'").all();

      expect((tables as TableInfo[]).map((t) => t.name)).toContain('oncall_schedule');
      expect((tables as TableInfo[]).map((t) => t.name)).toContain('oncall_schedule_overrides');

      // Run generation and verify real database interactions
      await runScheduleGeneration();

      // Check that actual data was inserted
      const scheduleCount = testDb.prepare('SELECT COUNT(*) as count FROM oncall_schedule').get() as { count: number };

      expect(scheduleCount.count).toBeGreaterThan(0);
    });
  });
});
