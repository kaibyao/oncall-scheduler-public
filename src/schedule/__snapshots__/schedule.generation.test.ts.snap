// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Schedule Generation (Characterization Tests) > runScheduleGeneration > should capture current generation algorithm behavior in snapshot > generation-algorithm-baseline 1`] = `
[
  {
    "date": "2025-07-28",
    "engineer_email": "dave.cowart@company.com",
    "rotation": "AM",
  },
  {
    "date": "2025-07-28",
    "engineer_email": "alex@company.com",
    "rotation": "Core",
  },
  {
    "date": "2025-07-28",
    "engineer_email": "eng.02@company.com",
    "rotation": "PM",
  },
  {
    "date": "2025-07-29",
    "engineer_email": "dave.cowart@company.com",
    "rotation": "AM",
  },
  {
    "date": "2025-07-29",
    "engineer_email": "alex@company.com",
    "rotation": "Core",
  },
  {
    "date": "2025-07-29",
    "engineer_email": "eng.02@company.com",
    "rotation": "PM",
  },
  {
    "date": "2025-07-30",
    "engineer_email": "dave.cowart@company.com",
    "rotation": "AM",
  },
  {
    "date": "2025-07-30",
    "engineer_email": "alex@company.com",
    "rotation": "Core",
  },
  {
    "date": "2025-07-30",
    "engineer_email": "eng.02@company.com",
    "rotation": "PM",
  },
  {
    "date": "2025-07-31",
    "engineer_email": "dave.cowart@company.com",
    "rotation": "AM",
  },
  {
    "date": "2025-07-31",
    "engineer_email": "alex@company.com",
    "rotation": "Core",
  },
  {
    "date": "2025-07-31",
    "engineer_email": "eng.02@company.com",
    "rotation": "PM",
  },
  {
    "date": "2025-08-01",
    "engineer_email": "dave.cowart@company.com",
    "rotation": "AM",
  },
  {
    "date": "2025-08-01",
    "engineer_email": "alex@company.com",
    "rotation": "Core",
  },
  {
    "date": "2025-08-01",
    "engineer_email": "eng.02@company.com",
    "rotation": "PM",
  },
  {
    "date": "2025-08-04",
    "engineer_email": "jeffrey@company.com",
    "rotation": "AM",
  },
  {
    "date": "2025-08-04",
    "engineer_email": "jason@company.com",
    "rotation": "Core",
  },
  {
    "date": "2025-08-04",
    "engineer_email": "frances.jurek@company.com",
    "rotation": "PM",
  },
  {
    "date": "2025-08-05",
    "engineer_email": "jeffrey@company.com",
    "rotation": "AM",
  },
  {
    "date": "2025-08-05",
    "engineer_email": "jason@company.com",
    "rotation": "Core",
  },
  {
    "date": "2025-08-05",
    "engineer_email": "frances.jurek@company.com",
    "rotation": "PM",
  },
  {
    "date": "2025-08-06",
    "engineer_email": "jeffrey@company.com",
    "rotation": "AM",
  },
  {
    "date": "2025-08-06",
    "engineer_email": "jason@company.com",
    "rotation": "Core",
  },
  {
    "date": "2025-08-06",
    "engineer_email": "frances.jurek@company.com",
    "rotation": "PM",
  },
  {
    "date": "2025-08-07",
    "engineer_email": "jeffrey@company.com",
    "rotation": "AM",
  },
  {
    "date": "2025-08-07",
    "engineer_email": "jason@company.com",
    "rotation": "Core",
  },
  {
    "date": "2025-08-07",
    "engineer_email": "frances.jurek@company.com",
    "rotation": "PM",
  },
  {
    "date": "2025-08-08",
    "engineer_email": "jeffrey@company.com",
    "rotation": "AM",
  },
  {
    "date": "2025-08-08",
    "engineer_email": "jason@company.com",
    "rotation": "Core",
  },
  {
    "date": "2025-08-08",
    "engineer_email": "frances.jurek@company.com",
    "rotation": "PM",
  },
]
`;
